File: sdk/js/src/async-client.ts
/**
 * PackCV 异步客户端（Node.js + 并发场景）
 * @module async-client
 */

import { errorFromStatus, PackCVError, NetworkError } from './exceptions';
import type {
  ExtractResponse,
  ExtractOptions,
  QAResponse,
  BatchItem,
  BatchResult,
  UsageStats,
  Tenant,
  TenantCreate,
  ApiKey,
  HealthStatus,
  Provider,
  Scenario,
  WebhookSubscription,
  WebhookDispatchRequest,
  WebhookDispatchResult,
} from './types';

export interface AsyncPackCVOptions {
  api_key: string;
  base_url?: string;
  timeout?: number;
  max_retries?: number;
  max_concurrent?: number;
}

const DEFAULT_BASE_URL = 'https://api.packcv-ocr.io';
const DEFAULT_TIMEOUT = 60000;

/** 异步 PackCV 客户端 */
export class AsyncPackCVClient {
  public readonly api_key: string;
  public readonly base_url: string;
  public readonly timeout: number;
  public readonly max_retries: number;
  public readonly max_concurrent: number;

  constructor(options: AsyncPackCVOptions) {
    this.api_key = options.api_key;
    this.base_url = (options.base_url ?? DEFAULT_BASE_URL).replace(/\/$/, '');
    this.timeout = options.timeout ?? DEFAULT_TIMEOUT;
    this.max_retries = options.max_retries ?? 3;
    this.max_concurrent = options.max_concurrent ?? 5;
  }

  private async _request<T>(method: string, path: string, body?: unknown): Promise<T> {
    const url = `${this.base_url}${path}`;
    // eslint-disable-next-line @typescript-eslint/no-implied-eval
    const nodeFetch: (url: string, init: { method: string; headers: Record<string, string>; body: string | null; signal: AbortSignal }) => Promise<{ status: number; json: () => Promise<unknown> }> = new Function('return require("node-fetch")')();
    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), this.timeout);

    let res;
    try {
      res = await nodeFetch(url, {
        method,
        headers: {
          'Authorization': `Bearer ${this.api_key}`,
          'Content-Type': 'application/json',
        },
        body: body !== undefined ? JSON.stringify(body) : null,
        signal: controller.signal,
      });
    } catch (err) {
      const msg = err instanceof Error ? err.message : String(err);
      if (msg.includes('abort')) {
        throw new PackCVError(`Request timeout after ${this.timeout}ms`, 408, 'timeout');
      }
      throw new NetworkError(msg);
    } finally {
      clearTimeout(timer);
    }

    const status = res.status;
    let data: unknown;
    try {
      data = await res.json();
    } catch {
      data = {};
    }

    if (status >= 200 && status < 300) {
      return data as T;
    }
    const dataObj = (data ?? {}) as { message?: string; error?: string };
    const errMsg = dataObj.message ?? dataObj.error ?? `HTTP ${status}`;
    throw errorFromStatus(status, errMsg, data as Record<string, unknown>);
  }

  async extract(file_url: string, options: ExtractOptions = {}): Promise<ExtractResponse> {
    return this._request<ExtractResponse>('POST', '/api/v1/extract', { file_url, ...options });
  }

  async qa(file_url: string, question: string, options: Omit<ExtractOptions, 'user_question'> = {}): Promise<QAResponse> {
    return this._request<QAResponse>('POST', '/api/v1/qa', { file_url, user_question: question, ...options });
  }

  /** 并发批量提取 */
  async batch_extract(items: BatchItem[], concurrency: number = this.max_concurrent): Promise<BatchResult> {
    const results: BatchResult['items'] = [];
    let succeeded = 0;
    let failed = 0;

    // 简单实现：分批
    for (let i = 0; i < items.length; i += concurrency) {
      const batch = items.slice(i, i + concurrency);
      const promises = batch.map(async (item) => {
        try {
          const data = await this.extract(item.file_url, { scenario: item.scenario, user_question: item.user_question });
          return { file_url: item.file_url, status: 'success' as const, data: data.structured_data };
        } catch (err) {
          return { file_url: item.file_url, status: 'failed' as const, error: err instanceof Error ? err.message : String(err) };
        }
      });
      const batchResults = await Promise.all(promises);
      for (const r of batchResults) {
        results.push(r);
        if (r.status === 'success') succeeded++;
        else failed++;
      }
    }

    return { total: items.length, succeeded, failed, items: results };
  }

  async list_scenarios(): Promise<Scenario[]> {
    const data = await this._request<Scenario[] | { scenarios?: Scenario[] }>('GET', '/api/v1/scenarios');
    if (Array.isArray(data)) return data;
    return (data?.scenarios ?? []) as Scenario[];
  }

  async list_tenants(): Promise<Tenant[]> {
    const data = await this._request<Tenant[] | { tenants?: Tenant[] }>('GET', '/api/v1/admin/tenants');
    if (Array.isArray(data)) return data;
    return (data?.tenants ?? []) as Tenant[];
  }

  async create_tenant(data: TenantCreate): Promise<Tenant> {
    return this._request<Tenant>('POST', '/api/v1/admin/tenants', data);
  }

  async get_tenant(tenant_id: string): Promise<Tenant> {
    return this._request<Tenant>('GET', `/api/v1/admin/tenants/${tenant_id}`);
  }

  async delete_tenant(tenant_id: string): Promise<void> {
    await this._request('DELETE', `/api/v1/admin/tenants/${tenant_id}`);
  }

  async list_api_keys(tenant_id: string): Promise<ApiKey[]> {
    const data = await this._request<ApiKey[] | { keys?: ApiKey[] }>('GET', `/api/v1/admin/tenants/${tenant_id}/api-keys`);
    if (Array.isArray(data)) return data;
    return (data?.keys ?? []) as ApiKey[];
  }

  async create_api_key(tenant_id: string, name: string, scopes: string[] = ['read', 'write']): Promise<ApiKey> {
    return this._request<ApiKey>('POST', `/api/v1/admin/tenants/${tenant_id}/api-keys`, { name, scopes });
  }

  async get_usage(tenant_id: string, period?: string): Promise<UsageStats> {
    const q = period ? `?period=${period}` : '';
    return this._request<UsageStats>('GET', `/api/v1/billing/usage/${tenant_id}${q}`);
  }

  async health(): Promise<HealthStatus> {
    return this._request<HealthStatus>('GET', '/api/v1/system/health');
  }

  async list_providers(): Promise<Provider[]> {
    const data = await this._request<Provider[] | { providers?: Provider[] }>('GET', '/providers');
    if (Array.isArray(data)) return data;
    return (data?.providers ?? []) as Provider[];
  }

  async subscribe_webhook(data: WebhookSubscription): Promise<WebhookSubscription> {
    return this._request<WebhookSubscription>('POST', '/webhooks/subscribe', data);
  }

  async dispatch_webhook(data: WebhookDispatchRequest): Promise<WebhookDispatchResult> {
    return this._request<WebhookDispatchResult>('POST', '/webhooks/dispatch', data);
  }
}

File: sdk/python/packcv/async_client.py
"""
PackCV-OCR 异步客户端（asyncio + httpx）
"""
import asyncio
import os
from typing import Any, Dict, List, Optional

import httpx

from packcv.client import PackCVClient
from packcv.exceptions import (
    APIError,
    AuthenticationError,
    PackCVError,
    QuotaExceededError,
    RateLimitError,
)
from packcv.types import (
    ExtractRequest,
    ExtractResponse,
    QAResponse,
    ScenarioInfo,
    TenantInfo,
    UsageInfo,
)


class AsyncPackCVClient:
    """PackCV-OCR 异步客户端

    Examples:
        >>> async with AsyncPackCVClient(api_key="pk_xxx") as client:
        ...     result = await client.extract_simple("https://example.com/id.jpg")
    """

    def __init__(
        self,
        api_key: str,
        api_secret: Optional[str] = None,
        base_url: Optional[str] = None,
        timeout: float = PackCVClient.DEFAULT_TIMEOUT,
        max_retries: int = PackCVClient.MAX_RETRIES,
        max_concurrency: int = 10,
    ):
        self.api_key = api_key
        self.api_secret = api_secret
        self.base_url = base_url or os.getenv("PACKCV_BASE_URL", PackCVClient.DEFAULT_BASE_URL).rstrip("/")
        self.timeout = timeout
        self.max_retries = max_retries
        self._semaphore = asyncio.Semaphore(max_concurrency)
        self._client: Optional[httpx.AsyncClient] = None

    def _default_headers(self) -> Dict[str, str]:
        h = {"User-Agent": "packcv-ocr-python-async/1.0.0", "Content-Type": "application/json"}
        h["X-API-Key"] = self.api_key
        if self.api_secret:
            h["X-API-Secret"] = self.api_secret
        return h

    async def __aenter__(self):
        self._client = httpx.AsyncClient(
            base_url=self.base_url,
            timeout=self.timeout,
            headers=self._default_headers(),
        )
        return self

    async def __aexit__(self, *args):
        if self._client:
            await self._client.aclose()

    async def _request(self, method: str, path: str, **kwargs) -> Dict[str, Any]:
        if not self._client:
            raise PackCVError("客户端未初始化，请使用 async with 语法")
        url = path if path.startswith("http") else f"{self.base_url}{path}"
        last_err: Optional[Exception] = None

        for attempt in range(self.max_retries + 1):
            try:
                async with self._semaphore:
                    resp = await self._client.request(method, url, **kwargs)
                self._raise_for_status(resp)
                if resp.status_code == 204 or not resp.content:
                    return {}
                return resp.json()
            except (httpx.TimeoutException, httpx.NetworkError) as e:
                last_err = e
                if attempt < self.max_retries:
                    await asyncio.sleep(2 ** attempt)
                    continue
            except RateLimitError as e:
                if e.retry_after and attempt < self.max_retries:
                    await asyncio.sleep(e.retry_after)
                    continue
                raise
        raise APIError(f"请求失败: {last_err}")

    def _raise_for_status(self, resp: httpx.Response):
        if resp.status_code < 400:
            return
        try:
            body = resp.json()
        except Exception:
            body = {"detail": resp.text}
        msg = body.get("detail") or body.get("message") or body.get("error") or resp.text
        if resp.status_code == 401:
            raise AuthenticationError(msg, status_code=401, response=body)
        if resp.status_code == 429:
            retry_after = int(resp.headers.get("Retry-After", "1"))
            raise RateLimitError(msg, retry_after=retry_after, status_code=429, response=body)
        if resp.status_code == 402:
            raise QuotaExceededError(msg, status_code=402, response=body)
        raise APIError(msg, status_code=resp.status_code, response=body)

    # ============= 业务接口 =============

    async def health(self) -> Dict[str, Any]:
        return await self._request("GET", "/api/v1/health")

    async def extract(self, request: ExtractRequest) -> ExtractResponse:
        data = await self._request("POST", "/api/v1/extract", json=request.model_dump(mode="json"))
        return ExtractResponse(**data)

    async def extract_simple(self, file_url: str, scenario: Optional[str] = None, user_question: str = "") -> ExtractResponse:
        return await self.extract(ExtractRequest(file_url=file_url, scenario=scenario, user_question=user_question))

    async def qa(self, question: str, file_url: Optional[str] = None) -> QAResponse:
        payload: Dict[str, Any] = {"question": question}
        if file_url:
            payload["file_url"] = file_url
        data = await self._request("POST", "/api/v1/qa", json=payload)
        return QAResponse(**data)

    async def batch_extract(self, file_urls: List[str], scenario: Optional[str] = None) -> List[ExtractResponse]:
        data = await self._request("POST", "/api/v1/batch", json={"file_urls": file_urls, "scenario": scenario})
        return [ExtractResponse(**r) for r in data.get("results", [])]

    async def list_scenarios(self) -> List[ScenarioInfo]:
        data = await self._request("GET", "/api/v1/scenarios")
        if isinstance(data, list):
            return [ScenarioInfo(**s) for s in data]
        return [ScenarioInfo(**s) for s in data.get("scenarios", data.get("data", []))]

    async def get_usage(self, year_month: Optional[str] = None) -> UsageInfo:
        params = {"year_month": year_month} if year_month else {}
        data = await self._request("GET", "/api/v1/usage", params=params)
        return UsageInfo(**data)

    async def me(self) -> TenantInfo:
        data = await self._request("GET", "/api/v1/me")
        return TenantInfo(**data)
